### ISSUE TITLE
Gradients on android

### ISSUE THREAD

**Issue by sonicdebris** [memononen/nanovg#178]
I'm trying to use nanovg on android using ndk, putting together a test app to share back (really appreciated what you did here). I was able to compile and launch the renderDemo function. It basically works, but not everything. I am getting an egl surface and context in c++ from a java surface, and drawing in it using nanovg:

```
const EGLint configAttribs[] = { 
        EGL_RENDERABLE_TYPE,EGL_OPENGL_ES2_BIT,    
        EGL_SURFACE_TYPE,EGL_WINDOW_BIT,
        EGL_BLUE_SIZE,8, EGL_GREEN_SIZE,8, EGL_RED_SIZE,8, EGL_ALPHA_SIZE,8,
        EGL_STENCIL_SIZE,8, EGL_NONE };
const EGLint contextAttribs[] = {EGL_CONTEXT_CLIENT_VERSION,2, EGL_NONE };
const EGLint surfaceAttribs[] = {EGL_RENDER_BUFFER,EGL_BACK_BUFFER, EGL_NONE };

EGLint numConfigs, format;
EGLConfig config;
EGLDisplay display = eglGetDisplay(EGL_DEFAULT_DISPLAY);
eglInitialize(display, NULL, NULL);
eglChooseConfig(display, configAttribs, &config, 1, &numConfigs);
EGLContext context = eglCreateContext(display, config, EGL_NO_CONTEXT, contextAttribs);
eglGetConfigAttrib(display, config, EGL_NATIVE_VISUAL_ID, &format);
ANativeWindow* window = ANativeWindow_fromSurface(e, jsurface);
ANativeWindow_setBuffersGeometry(window, 0, 0, format);
EGLSurface surface = eglCreateWindowSurface(display, config, window, surfaceAttribs);
eglMakeCurrent(display, surface, surface, context);
int width,height;
eglQuerySurface(display, surface, EGL_WIDTH, &width);
eglQuerySurface(display, surface, EGL_HEIGHT, &height);
glViewport(0, 0, width,height);

NVGcontext* vg = nvgCreateGLES2(NVG_ANTIALIAS | NVG_STENCIL_STROKES);
float pxRatio = 1.0;
glClearColor(0.3f, 0.3f, 0.32f, 1.0f);
glClear(GL_COLOR_BUFFER_BIT|GL_DEPTH_BUFFER_BIT|GL_STENCIL_BUFFER_BIT);
glEnable(GL_BLEND);
glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
glEnable(GL_CULL_FACE);
glDisable(GL_DEPTH_TEST);

nvgBeginFrame(vg, width, height, pxRatio);
renderDemo(vg, 200,200, width,height,5,0,0);
nvgEndFrame(vg);

eglSwapBuffers(display, surface);
```

And I get this:
![nanovg_android_test](https://cloud.githubusercontent.com/assets/1275794/4514402/03155fa0-4b6b-11e4-920e-154d0b7de00f.png)

I am ignoring the text and images absence for now (it's just a matter of loading the resources using android's APIs). I'm concerned about gradients, as you can see from the color wheel.

I'm totally new to opengl, so I don't know what that could be, I'm searching but maybe it's something trivial that I overlooked. Any idea/hint?

**Comment by sonicdebris** [comment]
I just tried again the test app I made, after downloading a "fresh" zip of the git repo. The gradient is now working!

I will perform further tests and I'd like to find a way to publish a polished test app (time to learn how to use git I suppose...), but since this might not happen very soon here's a zip with the current state of the app: https://www.dropbox.com/s/saspblk3yxanw60/nanoVGTest_android.zip?dl=0

I tested the app with a motorola moto g (stock android 4.4), and a Lg 4X (cyanogen 4.4).

**Comment by rjklindsay** [comment]
As of July 2015, Linear gradients do not work on Android. (Tested on Samsung Galaxy Note II)
The entire stroke just uses the second colour specified in nvgLinearGradient();
However, it does work correctly on desktop Ubuntu 14.04, with an GLES 2 rendering context.
Switching from mediump to highp, or downloading the latest source had no effect for me.

My current workaround is to use one edge of a nvgBoxGradient(), instead of nvgLinearGradient().

**Comment by memononen** [comment]
All gradients are done using box gradient, maybe choosing smaller `large`could fix it?
https://github.com/memononen/nanovg/blob/master/src/nanovg.c

The precision issues is quite sad... if someone has good tips how to solve it (maybe change the data), I'd like to hear about it.

**Comment by rjklindsay** [comment]
You're right.  If I reduce "const float large" from 1e5 to 1e4, the gradient contains 3 distinct colours.
Reducing if further, to 1e3, gives a smooth gradient.  Thanks for the tip.
